src/runtime/storageNG/storage-proxy.ts

```diff
@@ -21,19 +29,47 @@ export class StorageProxy<T extends CRDTTypeRecord> {
   private crdt: CRDTModel<T>;
   private id: number;
   private store: ActiveStore<T>;
+  private listenerAttached = false;
+  private keepSynced = false;
+  private synchronized = SyncState.none;
+  private readonly scheduler: StorageProxyScheduler<T>;
 
-  constructor(crdt: CRDTModel<T>, store: ActiveStore<T>) {
+  constructor(
+      crdt: CRDTModel<T>,
+      store: ActiveStore<T>,
+      scheduler: StorageProxyScheduler<T>) {
     this.crdt = crdt;
-    this.registerWithStore(store);
-  }
-
-  registerWithStore(store: ActiveStore<T>) {
-    this.id = store.on(x => this.onMessage(x));
     this.store = store;
+    this.scheduler = scheduler;
   }
 
-  registerHandle(h: Handle<T>): VersionMap {
-    this.handles.push(h);
+  registerHandle(handle: Handle<T>): VersionMap {
+    this.handles.push(handle);
+
+    // Attach an event listener to the backing store when the first readable handle is registered.
+    if (!this.listenerAttached) {
+      this.id = this.store.on(x => this.onMessage(x));
+      this.listenerAttached = true;
+    }
+
+    // Change to synchronized mode as soon as we get any handle configured with keepSynced and send
+    // a request to get the full model (once).
+    // TODO: drop back to non-sync mode if all handles re-configure to !keepSynced.
+    if (handle.options.keepSynced) {
+      if (!this.keepSynced) {
+        this.synchronizeModel().catch(e => {
+          console.error('Error during SyncRequest', e);
```

Comment:
Not sure this is the right way to handle errors here. Logging can be pretty easily ignored, and if we do get errors in model syncs we'd probably want to know as soon as possible. I think this should just let the exception bubble (or, rather, use something like the old store's `reportExceptionInHost()` call to report this on the host side).

This can be addressed in a different CL, if you prefer.

Reply:
Added a TODO to address that.